debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordState.java

```diff
@@ -143,6 +164,55 @@ public R apply(final R record) {
         }
     }
 
+    private R addSourceFields(String[] addSourceFields, R originalRecord, R unwrappedRecord) {
+        // Return if no source fields to add
+        if(addSourceFields == null) {
+            return unwrappedRecord;
+        }
+        
+        final Struct value = requireStruct(unwrappedRecord.value(), PURPOSE);
+        Struct source = ((Struct) originalRecord.value()).getStruct("source");
+        
+        // Get the updated schema from the cache, or create and cache if it doesn't exist
+        Schema updatedSchema = schemaUpdateCache.get(value.schema());
+        if (updatedSchema == null) {
+            updatedSchema = makeUpdatedSchema(value.schema(), addSourceFields);
+            schemaUpdateCache.put(value.schema(), updatedSchema);
+        }   
+
+        // Create the updated struct
+        final Struct updatedValue = new Struct(updatedSchema);
+        for (org.apache.kafka.connect.data.Field field : value.schema().fields()) {
+            updatedValue.put(field.name(), value.get(field));
+        }
+        for(String sourceField : addSourceFields) {
+            String fieldValue = source.schema().field(sourceField) == null ? "" : source.get(sourceField).toString();
```

Comment:
See below, I think it'd be better to propagate the original type. Plus, some fields are optional, so the `toString()` might pose a potential NPE?

Reply:
Agreed... Fixed to keep the original type, and handle optional fields w/ test